File: 0000_0099/leetcode-no20.cpp
#include <string>
#include <iostream>
#include <vector>
#include <map>
using namespace std;
class Solution {
public:
    bool isValid(string s) {
        vector<char> alist;
        string first = "([{";
        string second = "}])";
        map<char,char> dic = {{'(',')'},{'[',']'},{'{','}'}};
        char k;
        for (char i : s){
            if (first.find(i) != first.size()){
                alist.push_back(i);
            }
            printf("%c\n",i);
            if (second.find(i) != second.size()){
                try{
                    k = alist[alist.size()];
                } catch (exception){
                    return false;
                }
                if (i != dic[k]){
                    return false;
                }
            }
        }
        printf("asdf");
        if (alist.size() == 0){
            return true;
        } else {
            return false;
        }
    }
};

int main(){
    Solution t;
    printf(t.isValid("()[]{}") ? "yes" : "no");
}

File: 0000_0099/leetcode-no20.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        alist = []
        dic = {"(":")","[":"]","{":"}"}
        for i in s:
            if i in "([{" :
                alist.append(i)

            if i in ")]}" :
                try:
                    k = alist.pop()
                except:
                    return False
                if i != dic[k]:
                    return False
        if alist == []:
            return True
        else :
            return False

t=Solution
if t.isValid(6,"()[]{}"):
    print("yes")

